src/include/sof/lib/perf_cnt.h

```diff
@@ -0,0 +1,69 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2019 Intel Corporation. All rights reserved.
+ *
+ * Author: Marcin Maka <marcin.maka@linux.intel.com>
+ */
+
+/**
+ * \file include/sof/lib/perf_cnt.h
+ * \brief Simple performance counters
+ * \author Marcin Maka <marcin.maka@linux.intel.com>
+ */
+
+#ifndef __SOF_LIB_PERF_CNT_H__
+#define __SOF_LIB_PERF_CNT_H__
+
+#include <sof/drivers/timer.h>
+
+struct perf_cnt_data {
+	uint64_t plat_ts;
+	uint64_t cpu_ts;
+	uint64_t plat_delta_last;
+	uint64_t plat_delta_peak;
+	uint64_t cpu_delta_last;
+	uint64_t cpu_delta_peak;
+};
+
+#if CONFIG_PERF_CNT
+
+#define perf_cnt_trace(tclass, pcd) \
+		trace_event(tclass, "perf plat last %lu peak %lu cpu last %lu, peak %lu", \
+			    (pcd)->plat_delta_last,	\
+			    (pcd)->plat_delta_peak,	\
+			    (pcd)->cpu_delta_last,	\
+			    (pcd)->cpu_delta_peak)
+
+#define perf_cnt_clear(pcd) memset((pcd), 0, sizeof(struct perf_cnt_data))
+
+#define perf_cnt_init(pcd) do {						\
+		(pcd)->plat_ts = platform_timer_get(timer_get());	\
+		(pcd)->cpu_ts = arch_timer_get_system(cpu_timer_get());	\
+	} while (0)
+
+#define perf_cnt_stamp(tclass, pcd, log_peak) do {			  \
```

Comment:
Why so massive macro? Cannot we have static inline function here? Maybe even hide impl in .c file (you can also move out traces from header if you do that), but idk what affect on performance measuring would extra call have.

Reply:
There is a trace macro inside, and it contains static initializer. I am refactoring this code a little bit since the macros are not very readable atm indeed (were shorter initially).

Reply:
There are other issues with inlines since symbols (also compiled in conditionally) must exist before the code is eventually optimized. So I would keep it as is until a final widely adopted and unconditionally compiled framework is developed.